File: 13.roman-to-integer.java
/*
 * @lc app=leetcode id=13 lang=java
 *
 * [13] Roman to Integer
 */

// @lc code=start
class Solution {
    public int romanToInt(String s) {
        int sum = 0;
        int i = 0;
        int l = s.length();
        while (i < l) {
            switch (s.charAt(i)) {
                case 'I':
                    if (i != l - 1) {
                        if (s.charAt(i+1) =='V') {
                            sum += 4;
                            i++;
                        } else if (s.charAt(i+1) == 'X') {
                            sum += 9;
                            i++;
                        } else {
                            sum += 1;
                        }
                    } else {
                        sum += 1;
                    }   
                    i++;
                    break;
                
                case 'V':
                    sum += 5;
                    i++;
                    break;
                
                case 'X':
                    if (i != l - 1) {
                        if (s.charAt(i+1) =='L') {
                            sum += 40;
                            i++;
                        } else if (s.charAt(i+1) == 'C') {
                            sum += 90;
                            i++;
                        } else {
                            sum += 10;
                        }
                    } else {
                        sum += 10;
                    }
                    i++;
                    break;
                
                case 'L':
                    sum += 50;
                    i++;
                    break;                  
                
                case 'C':
                    if (i != l - 1) {
                        if (s.charAt(i+1) =='D') {
                            sum += 400;
                            i++;
                        } else if (s.charAt(i+1) == 'M') {
                            sum += 900;
                            i++;
                        } else {
                            sum += 100;
                        }
                    } else {
                        sum += 100;
                    }
                    i++;
                    break;
                
                case 'D':
                    sum += 500;
                    i++;
                    break;
                
                case 'M':
                    sum += 1000;
                    i++;
                    break;

            }
        }


        return sum;
    }
}
// @lc code=end

File: 13.roman-to-integer.py
#
# @lc app=leetcode id=13 lang=python3
#
# [13] Roman to Integer
#

# @lc code=start
class Solution:
    def romanToInt(self, s: str) -> int:
        d = {
            "I": 1,
            "IV": 4,
            "V": 5,
            "IX": 9,
            "X": 10,
            "XL": 40,
            "L": 50,
            "XC": 90,
            "C": 100,
            "CD": 400,
            "D": 500,
            "CM": 900,
            "M": 1000,
        }
        total = 0
        index = 0
        len_s = len(s)
        while index + 1 < len_s:
            if s[index : index + 2] in d:
                total += d[s[index : index + 2]]
                index += 2
                print(f"index + 2 = {index}")
            else:
                total += d[s[index]]
                index += 1
                print(f"index + 1 = {index}")

        if index + 1 == len_s:
            total += d[s[index]]

        return total


# @lc code=end
